In [ ]:
import tensorflow as tf
import numpy as np                                               # 1)
import keras   
import pandas as pd                                                  # 2)
from keras.models import Sequential                              # 3)
from keras.layers import Dense, Dropout, Activation, Flatten     # 4)
from keras.datasets import cifar10                               # 5)
from keras.layers import Conv2D, MaxPooling2D                    # 6)
from keras.preprocessing.image import ImageDataGenerator         # 7)
import matplotlib.pyplot as plt                                  # 8)
from keras.utils import np_utils                                 # 9)
from keras.layers.normalization import BatchNormalization


In [ ]:
cd /content/drive/MyDrive/Colab\ Notebooks/CS680\ Kaggle/

/content/drive/MyDrive/Colab Notebooks/CS680 Kaggle


In [ ]:
import zipfile
with zipfile.ZipFile("train.csv.zip", 'r') as zip_ref:
    zip_ref.extractall("./")

In [ ]:
import zipfile
with zipfile.ZipFile("test.csv.zip", 'r') as zip_ref:
    zip_ref.extractall("./")

In [ ]:
import zipfile
with zipfile.ZipFile("suffled-images.zip", 'r') as zip_ref:
    zip_ref.extractall("./")

KeyboardInterrupt: ignored

Sort images in ascending order (same as labels):

In [ ]:
import os
dir=os.listdir('shuffled-images/')
from tkinter import Tcl
dir_sort=Tcl().call('lsort', '-dict', dir)
print(dir_sort[:10])
len(dir_sort)


('1163.jpg', '1164.jpg', '1165.jpg', '1525.jpg', '1526.jpg', '1528.jpg', '1529.jpg', '1530.jpg', '1531.jpg', '1532.jpg')


43255

Total of 43255 images:

"loaded_images" is list of images in tensor format :

In [ ]:
train=pd.read_csv('train.csv',header=0)
print(train.columns)
train.head#21627

Index(['id', 'category', 'gender', 'baseColour', 'season', 'usage',
       'noisyTextDescription'],
      dtype='object')


<bound method NDFrame.head of           id  ...                               noisyTextDescription
0      36274  ...  Femella Women Ankle-Length Grey AQ-S800WD-1EVD...
1      15129  ...              Converse Unisex Casual Skirts Slipper
2      58976  ...                          Velia Women Acetone Kurta
3      32922  ...                  Enroute Men Leather Brown Sandals
4      29561  ...              Aneri Exclusive Anu Pink Inspirartion
...      ...  ...                                                ...
21622   5435  ...                    SHE-3802D-1ADR Men Black Briefs
21623  27880  ...                 Adipower & Taylor Men Brown Quince
21624  38385  ...                             Rocia Women R348 Flats
21625  34853  ...                  Fastrack Men Black Amethyst Watch
21626  41407  ...             Arrow Ghicha Men White Slim Surf Shirt

[21627 rows x 7 columns]>

In [ ]:
test=pd.read_csv('test.csv',header=0)
print(test.columns)
test.head#21628

Index(['id', 'gender', 'baseColour', 'season', 'usage',
       'noisyTextDescription'],
      dtype='object')


<bound method NDFrame.head of           id  gender  ...   usage                               noisyTextDescription
0      26266     Men  ...  Casual                  Chromozome Men Black Fashion Vest
1      22134   Women  ...  Casual                   Elle Women Green Color Clash Top
2      28358   Women  ...  Casual               Baggit Women Chotu Mayur Black Palms
3      15554     Men  ...  Casual  Greensboro Colors Of Ap Men Latnam Black Casua...
4      53408   Women  ...  Casual                    Fish White Smooth daddy (SH100)
...      ...     ...  ...     ...                                                ...
21623  39737     Men  ...  Casual                Park ClubLife 602 Helium Sunglasses
21624  57477   Women  ...  Casual              Vero Moda Women Black Rey Shift Dress
21625  22312     Men  ...  Sports                   Nike Men Hell With White Tshirts
21626  54105   Women  ...  Casual       Lakme Holi to Five Day Apricot Nectar Lip K9
21627  14080  Unisex  ...  Casual  

# "loaded_images_train": Create a List of tensor coresponding to columen "ID" in data frame "Train":

In [ ]:
from os import listdir
from matplotlib import image
# load all images in a directory
loaded_images_train = list()
for filename in train['id']:
    #print(filename)
  # load image
    img_data = image.imread('shuffled-images/' + str(filename) + ".jpg")
    # store loaded image
    loaded_images_train.append(img_data)
   # print('> loaded %s %s' % (filename, img_data.shape))

In [ ]:
np.array(loaded_images_train).shape

(21627, 80, 60, 3)

# "loaded_images_test": Create a List of tensor coresponding to columen "ID" in data frame "Test":

In [ ]:
from os import listdir
from matplotlib import image
# load all images in a directory
loaded_images_test = list()
for filename in test['id']:
    #print(filename)
  # load image
    img_data = image.imread('shuffled-images/' + str(filename) + ".jpg")
    # store loaded image
    loaded_images_test.append(img_data)
   # print('> loaded %s %s' % (filename, img_data.shape))

Construct a data frame called "data_image": column1 is ID of image, column2 is tensor of each image.

result "data_image" is a list of tuples

 Construct a data frame corresponding to training data:


Construct a data frame corresponding to test data:




convert same column to the same data type before merging two data frames based on this column:

"train_data" is data frame with "tensor" as new column:

In [ ]:
train_data = pd.merge(left=data_image, right=train, left_on='id', right_on='id')
train_data

id  ...                               noisyTextDescription
0       1164  ...  tenkay tv Graphiti My Fab Eyewear Grey Sunglasses
1       1165  ...                           Kiara Women Pink Handbag
2       1526  ...           Puma VH220-C3 Even Keel Casual Pinstripe
3       1529  ...                            Enamor Rose Wo Lips Bra
4       1530  ...            F Khadi Anti-Ageing Hustle Falcon Shoes
...      ...  ...                                                ...
21622  59987  ...                   Carlton London Nayak Filac Heels
21623  59989  ...                         Murcia Women Brown Handbag
21624  59991  ...  Flower Colors of Cricket Men Light Winter Purp...
21625  59994  ...                       HM PFFL01 Multi Khakhi Flats
21626  59996  ...            Colorbar Toast to Pink Nail Lacquer 049

[21627 rows x 8 columns]

In [ ]:
test_data = pd.merge(left=data_image, right=test, left_on='id', right_on='id')
test_data

id  ...                            noisyTextDescription
0       1163  ...       CASIO Men Reglur Dial Analogue Watch A555
1       1525  ...       014 Barati Sports Socks 1 Pair Grey Socks
2       1528  ...       Disney Carlton Girl's Check Pink Kidswear
3       1531  ...                    Remanika Women Navy Blue Top
4       1533  ...                         A629 Men Yellow T-shirt
...      ...  ...                                             ...
21623  59993  ...        Scullers Men Green & Ace Striped T-shirt
21624  59995  ...  Nike Women Ballista Else White Sports Heritage
21625  59998  ...                    Miami Blues Women Sunglasses
21626  59999  ...      Allen Solly Kids Girls Polka Green T-shirt
21627  60000  ...               Lee Cooper Men Black Formal Shoes

[21628 rows x 7 columns]

 1) CNN to classify only based on images:



*Initialize* parameters:

In [ ]:
batch_size = 256      #batch size
num_classes = 27      #number of classes
epochs = 100          #epoch size

In [ ]:
label=train['category'].tolist()
label

['Scarves',
 'Flip Flops',
 'Topwear',
 'Sandal',
 'Topwear',
 'Bags',
 'Socks',
 'Topwear',
 'Topwear',
 'Shoes',
 'Scarves',
 'Bags',
 'Topwear',
 'Watches',
 'Topwear',
 'Topwear',
 'Watches',
 'Topwear',
 'Sandal',
 'Dress',
 'Shoes',
 'Topwear',
 'Shoes',
 'Headwear',
 'Topwear',
 'Jewellery',
 'Bottomwear',
 'Topwear',
 'Shoes',
 'Bags',
 'Bags',
 'Topwear',
 'Bags',
 'Innerwear',
 'Shoes',
 'Wallets',
 'Belts',
 'Sandal',
 'Saree',
 'Topwear',
 'Watches',
 'Topwear',
 'Topwear',
 'Jewellery',
 'Shoes',
 'Jewellery',
 'Wallets',
 'Shoes',
 'Topwear',
 'Sandal',
 'Topwear',
 'Innerwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Nails',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Shoes',
 'Topwear',
 'Topwear',
 'Socks',
 'Loungewear and Nightwear',
 'Topwear',
 'Shoes',
 'Topwear',
 'Shoes',
 'Shoes',
 'Sandal',
 'Topwear',
 'Topwear',
 'Lips',
 'Bags',
 'Shoes',
 'Dress',
 'Shoes',
 'Topwear',
 'Topwear',
 'Shoes',
 'Nails',
 'Shoes',
 'Topwear',
 'Inne

In [ ]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
label_list=train['category'].tolist()
le.fit(label_list)
transfomed_label =le.transform(label_list)
print(transfomed_label)
print(transfomed_label.shape)

[20  8 24 ... 21 26 24]
(21627,)


In [ ]:
list(le.inverse_transform(transfomed_label))

['Scarves',
 'Flip Flops',
 'Topwear',
 'Sandal',
 'Topwear',
 'Bags',
 'Socks',
 'Topwear',
 'Topwear',
 'Shoes',
 'Scarves',
 'Bags',
 'Topwear',
 'Watches',
 'Topwear',
 'Topwear',
 'Watches',
 'Topwear',
 'Sandal',
 'Dress',
 'Shoes',
 'Topwear',
 'Shoes',
 'Headwear',
 'Topwear',
 'Jewellery',
 'Bottomwear',
 'Topwear',
 'Shoes',
 'Bags',
 'Bags',
 'Topwear',
 'Bags',
 'Innerwear',
 'Shoes',
 'Wallets',
 'Belts',
 'Sandal',
 'Saree',
 'Topwear',
 'Watches',
 'Topwear',
 'Topwear',
 'Jewellery',
 'Shoes',
 'Jewellery',
 'Wallets',
 'Shoes',
 'Topwear',
 'Sandal',
 'Topwear',
 'Innerwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Nails',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Shoes',
 'Topwear',
 'Topwear',
 'Socks',
 'Loungewear and Nightwear',
 'Topwear',
 'Shoes',
 'Topwear',
 'Shoes',
 'Shoes',
 'Sandal',
 'Topwear',
 'Topwear',
 'Lips',
 'Bags',
 'Shoes',
 'Dress',
 'Shoes',
 'Topwear',
 'Topwear',
 'Shoes',
 'Nails',
 'Shoes',
 'Topwear',
 'Inne

In [ ]:
x_train=np.array(loaded_images_train)
y_train=np.array(transfomed_label)
x_test=np.array(loaded_images_test)
x_val=x_train[-4325:]
y_val=y_train[-4325:]
x_train=x_train[:-4325]
y_train=y_train[:-4325]
print("x_train is ",x_train.shape[0])
print("x_val is ",x_val.shape[0])
print("x_test is ",x_test.shape[0])


x_train is  17302
x_val is  4325
x_test is  21628


In [ ]:
y_train = np_utils.to_categorical(y_train, num_classes)#convert train labels to one hot encoding
y_val = np_utils.to_categorical(y_val, num_classes)  #convert test labels to one hot encoding

In [ ]:
y_train[:3]

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0.]], dtype=float32)

In [ ]:
y_train.shape

(17302, 27)

In [ ]:
y_val.shape

(4325, 27)

In [ ]:
#Typecast data samples to flot32. It is usefull when using GPU.
x_train = x_train.astype('float32')   
x_test = x_test.astype('float32')
x_val=x_val.astype('float32')
# Scale pixles of data samples between 0 and 1.
x_train /= 255
x_test /= 255
x_val /= 255

Need to convert string multiclass label into integers and then apply one-hot encoding: using SKlearn package to do this :

In [ ]:
# define a function to train with or without data augmentation:
def data_aug (save_best_model):
  if not data_augmentation:
      print('Not using data augmentation.')
      history = model.fit(x_train, y_train,verbose=1,
                          batch_size=batch_size,
                          epochs=epochs,
                          validation_data=(x_val,y_val),
                          shuffle=True,
                          callbacks=[save_best_model])

  # train with data augmentation
  else:
      print('Using real-time data augmentation.')
      # This will do preprocessing and realtime data augmentation:
      datagen = ImageDataGenerator(
          featurewise_center=False,  # set input mean to 0 over the dataset
          samplewise_center=False,  # set each sample mean to 0
          featurewise_std_normalization=False,  # divide inputs by std of the dataset
          samplewise_std_normalization=False,  # divide each input by its std
          zca_whitening=False,  # apply ZCA whitening
          zca_epsilon=1e-06,  # epsilon for ZCA whitening
          rotation_range=0,  # randomly rotate images in the range (degrees, 0 to 180)
          # randomly shift images horizontally (fraction of total width)
          width_shift_range=0.1,
          # randomly shift images vertically (fraction of total height)
          height_shift_range=0.1,
          shear_range=0.,  # set range for random shear
          zoom_range=0.,  # set range for random zoom
          channel_shift_range=0.,  # set range for random channel shifts
          # set mode for filling points outside the input boundaries
          fill_mode='nearest',
          cval=0.,  # value used for fill_mode = "constant"
          horizontal_flip=True,  # randomly flip images
          vertical_flip=False,  # randomly flip images
          # set rescaling factor (applied before any other transformation)
          rescale=None,
          # set function that will be applied on each input
          preprocessing_function=None,
          # image data format, either "channels_first" or "channels_last"
          data_format=None,
          # fraction of images reserved for validation (strictly between 0 and 1)
          validation_split=0.0)

      # Compute quantities required for feature-wise normalization
      # (std, mean, and principal components if ZCA whitening is applied).
      datagen.fit(x_train)

      # Fit the model on the batches generated by datagen.flow().
      history = model.fit(datagen.flow(x_train, y_train, batch_size=batch_size),verbose=1,
                          steps_per_epoch=math.ceil(x_train.shape[0]/batch_size),
                          epochs=epochs,
                          validation_data=(x_val,y_val),
                          callbacks=[save_best_model])
  return(history)   

In [ ]:
x_train.shape

(17302, 80, 60, 3)

In [ ]:
# libraries (do not import additional libraries)
import keras
from keras.datasets import cifar10
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential, load_model
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras.callbacks import ModelCheckpoint
import math
import matplotlib.pyplot as plt

# parameters for this script
batch_size = 32
num_classes = 27
epochs = 20
data_augmentation = True


# Define a convolutional neural network
model = Sequential()
model.add(Conv2D(32, (3, 3), padding='same',input_shape=x_train.shape[1:]))
model.add(Activation('relu'))
model.add(BatchNormalization())
model.add(Conv2D(32, (3, 3), padding='same'))
model.add(Activation('relu'))
model.add(BatchNormalization())
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Conv2D(64, (3, 3), padding='same'))
model.add(Activation('relu'))
model.add(BatchNormalization())
model.add(Conv2D(64, (3, 3), padding='same'))
model.add(Activation('relu'))
model.add(BatchNormalization())
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Conv2D(64, (3, 3), padding='same'))
model.add(Activation('relu'))
model.add(BatchNormalization())
model.add(Conv2D(64, (3, 3), padding='same'))
model.add(Activation('relu'))
model.add(BatchNormalization())
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Flatten())
model.add(Dense(512))
model.add(Activation('relu'))
model.add(BatchNormalization())

model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

# initiate RMSprop optimizer
opt = keras.optimizers.RMSprop(lr=0.0001, decay=1e-6)

# Compile the model before using it
model.compile(loss='categorical_crossentropy',
              optimizer=opt,
              metrics=['accuracy'])
print(model.summary())

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_17 (Conv2D)           (None, 80, 60, 32)        896       
_________________________________________________________________
activation_24 (Activation)   (None, 80, 60, 32)        0         
_________________________________________________________________
batch_normalization_4 (Batch (None, 80, 60, 32)        128       
_________________________________________________________________
conv2d_18 (Conv2D)           (None, 80, 60, 32)        9248      
_________________________________________________________________
activation_25 (Activation)   (None, 80, 60, 32)        0         
_________________________________________________________________
batch_normalization_5 (Batch (None, 80, 60, 32)        128       
_________________________________________________________________
max_pooling2d_8 (MaxPooling2 (None, 40, 30, 32)       

In [ ]:
x_train.shape

(17302, 80, 60, 3)

In [ ]:
y_train.shape

(17302, 27)

In [ ]:
x_val.shape

(4325, 80, 60, 3)

In [ ]:
# create a callback that will save the best model while training
save_best_model = ModelCheckpoint('best_model.h5', monitor='val_accuracy', mode='max', save_best_only=True, verbose=1)

# call function "data_aug" to train without data augmentation:
history=data_aug(save_best_model)


# Evaluate the best model saved (i.e., model with best validation accuracy) on the test set
saved_model = load_model('best_model.h5')


Using real-time data augmentation.
Epoch 1/20
540/541 [============================>.] - ETA: 0s - loss: 1.6153 - accuracy: 0.6051
Epoch 00001: val_accuracy improved from -inf to 0.75006, saving model to best_model.h5
541/541 [==============================] - 26s 48ms/step - loss: 1.6138 - accuracy: 0.6054 - val_loss: 1.0052 - val_accuracy: 0.7501
Epoch 2/20
540/541 [============================>.] - ETA: 0s - loss: 1.0853 - accuracy: 0.7145
Epoch 00002: val_accuracy improved from 0.75006 to 0.75514, saving model to best_model.h5
541/541 [==============================] - 26s 48ms/step - loss: 1.0848 - accuracy: 0.7147 - val_loss: 0.9136 - val_accuracy: 0.7551
Epoch 3/20
540/541 [============================>.] - ETA: 0s - loss: 0.9320 - accuracy: 0.7529
Epoch 00003: val_accuracy improved from 0.75514 to 0.80162, saving model to best_model.h5
541/541 [==============================] - 26s 48ms/step - loss: 0.9326 - accuracy: 0.7525 - val_loss: 0.7622 - val_accuracy: 0.8016
Epoch 4/20


Predict validation set and get accuracy:

In [ ]:
# This function of trained model predicts the label of its input data
prd = model.predict(x_test) 

In [ ]:
# These lines are used for converting one hot coding back to the original label form.
prd_y = np.argmax(prd, axis=1)


In [ ]:
print(prd_y)

[24 24  3 ...  9 14  2]


In [ ]:
label = list(le.inverse_transform(prd_y))
label



['Topwear',
 'Topwear',
 'Belts',
 'Shoes',
 'Innerwear',
 'Shoes',
 'Topwear',
 'Eyewear',
 'Watches',
 'Shoes',
 'Innerwear',
 'Topwear',
 'Ties',
 'Innerwear',
 'Socks',
 'Wallets',
 'Watches',
 'Watches',
 'Bottomwear',
 'Shoes',
 'Nails',
 'Topwear',
 'Bottomwear',
 'Belts',
 'Belts',
 'Topwear',
 'Bags',
 'Shoes',
 'Bags',
 'Topwear',
 'Topwear',
 'Bags',
 'Dress',
 'Bottomwear',
 'Topwear',
 'Eyewear',
 'Watches',
 'Wallets',
 'Watches',
 'Shoes',
 'Topwear',
 'Topwear',
 'Bottomwear',
 'Topwear',
 'Watches',
 'Topwear',
 'Flip Flops',
 'Topwear',
 'Topwear',
 'Bags',
 'Watches',
 'Saree',
 'Innerwear',
 'Topwear',
 'Topwear',
 'Shoes',
 'Wallets',
 'Bottomwear',
 'Innerwear',
 'Shoes',
 'Shoes',
 'Flip Flops',
 'Shoes',
 'Shoes',
 'Dress',
 'Shoes',
 'Topwear',
 'Bottomwear',
 'Topwear',
 'Watches',
 'Topwear',
 'Watches',
 'Topwear',
 'Flip Flops',
 'Watches',
 'Topwear',
 'Topwear',
 'Wallets',
 'Topwear',
 'Shoes',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Watches',


In [ ]:
predict=np.stack((test['id'],np.array(label)),axis=1)
predict


array([['26266', 'Topwear'],
       ['22134', 'Topwear'],
       ['28358', 'Belts'],
       ...,
       ['22312', 'Fragrance'],
       ['54105', 'Lips'],
       ['14080', 'Bags']], dtype='<U24')

In [ ]:
import csv

with open('submission.csv', mode='w') as csv_file:
    writer = csv.writer(csv_file)
    writer.writerow(['id','category'])
    writer.writerows(predict)

## 2) Use LTSM to model noisy text:

In [ ]:
train.columns

Index(['id', 'category', 'gender', 'baseColour', 'season', 'usage',
       'noisyTextDescription'],
      dtype='object')

In [ ]:
train[:3]

id    category  ...   usage                               noisyTextDescription
0  36274     Scarves  ...  Casual  Femella Women Ankle-Length Grey AQ-S800WD-1EVD...
1  15129  Flip Flops  ...  Casual              Converse Unisex Casual Skirts Slipper
2  58976     Topwear  ...  Ethnic                          Velia Women Acetone Kurta

[3 rows x 7 columns]

In [ ]:
import keras
import numpy as np
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import LSTM
from keras.callbacks import ModelCheckpoint


In [ ]:
max_features = 100 #This is highest frequency threshold for all indices. 
maxlen = 50  # cut texts after this number of words (among top max_features most common words)
batch_size = 32
num_classes = 27

In [ ]:
import re  # For preprocessing
from time import time  # To time our operations
from collections import defaultdict  # For word frequency
import spacy  # For preprocessing

In [ ]:
text=train['noisyTextDescription']
text_test=test['noisyTextDescription']

In [ ]:
text_combine=pd.concat([text,text_test])
len(text_combine)

43255

In [ ]:
text[:2]

0    Femella Women Ankle-Length Grey AQ-S800WD-1EVD...
1                Converse Unisex Casual Skirts Slipper
Name: noisyTextDescription, dtype: object

In [ ]:
from gensim.models.phrases import Phrases, Phraser

In [ ]:
#As Phrases() takes a list of list of words as input:
#Creates the relevant phrases from the list of sentences:
sent = [row.split() for row in text_combine]
phrases = Phrases(sent, progress_per=10000)
#The goal of Phraser() is to cut down memory consumption of Phrases(), by discarding model state not strictly needed for the bigram detection task:
sentences = phrases[sent]


/usr/local/lib/python3.6/dist-packages/gensim/models/phrases.py:598: UserWarning: For a faster implementation, use the gensim.models.phrases.Phraser class
  warnings.warn("For a faster implementation, use the gensim.models.phrases.Phraser class")


In [ ]:
#sent is list of words that contained in whole corpus
sent[0][:10]

['Femella', 'Women', 'Ankle-Length', 'Grey', 'AQ-S800WD-1EVDFAD170']

In [ ]:
word_freq = defaultdict(int)
for sent in sentences:
    for i in sent:
        word_freq[i] += 1
len(word_freq)

/usr/local/lib/python3.6/dist-packages/gensim/models/phrases.py:598: UserWarning: For a faster implementation, use the gensim.models.phrases.Phraser class
  warnings.warn("For a faster implementation, use the gensim.models.phrases.Phraser class")


10022

In [ ]:
sorted(word_freq, key=word_freq.get, reverse=True)[:10]

['Men',
 'Women',
 'Black',
 'White',
 'Blue',
 'T-shirt',
 'Printed',
 'Brown',
 'Grey',
 'Red']

In [ ]:
from gensim.models import Word2Vec

In [ ]:
import multiprocessing


In [ ]:
cores = multiprocessing.cpu_count() # Count the number of cores in a computer
cores

2

In [ ]:
w2v_model = Word2Vec(min_count=0,
                     window=2,
                     size=300,
                     sample=6e-5, 
                     alpha=0.03, 
                     min_alpha=0.0007, 
                     negative=20,
                     workers=cores-1)

In [ ]:
t = time()

w2v_model.build_vocab(sentences, progress_per=10000)

print('Time to build vocab: {} mins'.format(round((time() - t) / 60, 2)))

/usr/local/lib/python3.6/dist-packages/gensim/models/phrases.py:598: UserWarning: For a faster implementation, use the gensim.models.phrases.Phraser class
  warnings.warn("For a faster implementation, use the gensim.models.phrases.Phraser class")


Time to build vocab: 0.05 mins


In [ ]:
w2v_model.init_sims(replace=True)

In [ ]:
#most similiar words to "us"
w2v_model.wv.most_similar(positive=["male"])

/usr/local/lib/python3.6/dist-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('Trinity', 0.22353610396385193),
 ('Pandu', 0.21231773495674133),
 ('DL002', 0.20747897028923035),
 ('Kantha', 0.1899844855070114),
 ('A516', 0.1790388822555542),
 ('Henderson', 0.17743757367134094),
 ('LC', 0.17683859169483185),
 ('sports', 0.17656324803829193),
 ('NB9154YM05A', 0.1739329844713211),
 ('VH203-C3', 0.17391705513000488)]

In [ ]:
model_name = "product_content_clean"
w2v_model.save(model_name)

In [ ]:
model = Word2Vec.load("product_content_clean")

In [ ]:
word_vectors = model.wv.syn0 #array essentially holds raw word-vectors.these vectors are a 'projection layer' that can convert a one-hot encoding of a word into a dense embedding-vector of the right dimensionality.
word_vectors

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `syn0` (Attribute will be removed in 4.0.0, use self.wv.vectors instead).
  """Entry point for launching an IPython kernel.


array([[-0.06464316,  0.05755419,  0.00360025, ..., -0.04586885,
         0.0606902 , -0.02014844],
       [-0.06907661,  0.06359156,  0.00660026, ...,  0.08554636,
        -0.08100878, -0.02716447],
       [ 0.02433263,  0.00283197,  0.08360185, ...,  0.02033432,
         0.01145732, -0.00913811],
       ...,
       [ 0.02568038,  0.05542666,  0.078394  , ...,  0.05869623,
        -0.06105911,  0.01142298],
       [ 0.05022665, -0.06814919,  0.05471765, ..., -0.07755686,
         0.08617999,  0.04214151],
       [-0.08110292,  0.04873443, -0.05746552, ...,  0.03103261,
         0.06752995, -0.04114949]], dtype=float32)

In [ ]:
#word2index is dictionary of all words as key and value is frequency in corpus:
word2index = {token: token_index for token_index, token in enumerate(model.wv.index2word)}
"US" in word2index

False

In [ ]:
len(word2index)

10022

In [ ]:
print(model.wv.index2word[:10])
word="male"
print(model.wv.vocab.get(word).index)


['Men', 'Women', 'Black', 'White', 'Blue', 'T-shirt', 'Printed', 'Brown', 'Grey', 'Red']
9875


In [ ]:
sentences_index=[]
for news in sentences:
    lst=[]
    for word in news:
      if word in word2index:#if the word is in corpus
        if word2index[word]+1<max_features-1:#exclude extremely rare words
           lst.append(word2index[word]+1)
    sentences_index.append(lst)


/usr/local/lib/python3.6/dist-packages/gensim/models/phrases.py:598: UserWarning: For a faster implementation, use the gensim.models.phrases.Phraser class
  warnings.warn("For a faster implementation, use the gensim.models.phrases.Phraser class")


In [ ]:
len(sentences_index)

43255

In [ ]:
sentences_index[1][:10]

[24]

In [ ]:
train.shape

(21627, 7)

In [ ]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
label_list=train['category'].tolist()
le.fit(label_list)
transfomed_label =le.transform(label_list)
print(transfomed_label)
print(transfomed_label.shape)

[20  8 24 ... 21 26 24]
(21627,)


In [ ]:
transfomed_label

array([20,  8, 24, ..., 21, 26, 24])

In [ ]:
x_train_text=sentences_index[:21627]
x_test_text=sentences_index[21627:]
x_val_text=x_train_text[-4325:]
x_train_text=x_train_text[:-4325]
y_train_text=np.array(transfomed_label)
y_val_text=y_train_text[-4325:]
y_train_text=y_train_text[:-4325]

print(len(y_train_text))
print(len(x_train_text))
print(len(x_val_text))
print(len(y_val_text))

17302
17302
4325
4325


In [ ]:
y_train_text = np_utils.to_categorical(y_train_text, num_classes)#convert train labels to one hot encoding
y_val_text = np_utils.to_categorical(y_val_text, num_classes)  #convert test labels to one hot encoding

In [ ]:
# create a callback that will save the best model while training
checkpoint = ModelCheckpoint('best_model_text.h5', monitor='val_accuracy', mode='max', save_best_only=True, verbose=1)

In [ ]:
print('Pad sequences (samples x time)')
x_train_text = sequence.pad_sequences(x_train_text, maxlen=maxlen)
x_val_text = sequence.pad_sequences(x_val_text, maxlen=maxlen)
print('x_train shape:', x_train_text.shape)
print('x_test shape:', x_val_text.shape)

Pad sequences (samples x time)
x_train shape: (17302, 50)
x_test shape: (4325, 50)


In [ ]:
print('Pad sequences (samples x time)')
x_test_text = sequence.pad_sequences(x_test_text, maxlen=maxlen)

Pad sequences (samples x time)


In [ ]:
from keras.layers import Bidirectional


In [ ]:
print('Build model...')
model = Sequential()
model.add(Embedding(max_features, 10, input_length=maxlen))
model.add(Bidirectional(LSTM(10, return_sequences=True)))
model.add(Bidirectional(LSTM(10)))
model.add(Dense(num_classes, activation='softmax'))    


Build model...


Use plain LSTM model:

In [ ]:
model = Sequential()

#######################
#  Embedding layer[4] #
#######################
# - the model will take as input an integer matrix of size (batch_size, input_length).
# - the largest integer (i.e. word index) in the input should be no larger than (max_features-1) 
#   (vocabulary size).
# - now model.output_shape == (None, input_length, 128), where None is the batch dimension.
model.add(Embedding(max_features, 128, input_length=maxlen))
model.add(LSTM(128, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(num_classes, activation='softmax'))    

In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, 50, 128)           12800     
_________________________________________________________________
lstm (LSTM)                  (None, 128)               131584    
_________________________________________________________________
dense (Dense)                (None, 27)                3483      
Total params: 147,867
Trainable params: 147,867
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='rmsprop',
metrics=['accuracy'])

In [ ]:
# try using different optimizers and different optimizer configs
model.compile(loss='binary_crossentropy',
              optimizer='adam',
metrics=['accuracy'])

In [ ]:
print('Train...')
model.fit(x_train_text, y_train_text,
          batch_size=batch_size,
          epochs=15,callbacks=[checkpoint],
validation_data=(x_val_text, y_val_text))

Train...
Epoch 1/15
541/541 [==============================] - ETA: 0s - loss: 0.0788 - accuracy: 0.5921
Epoch 00001: val_accuracy improved from -inf to 0.70798, saving model to best_model_text.h5
541/541 [==============================] - 104s 192ms/step - loss: 0.0788 - accuracy: 0.5921 - val_loss: 0.0548 - val_accuracy: 0.7080
Epoch 2/15
541/541 [==============================] - ETA: 0s - loss: 0.0546 - accuracy: 0.7026
Epoch 00002: val_accuracy improved from 0.70798 to 0.71191, saving model to best_model_text.h5
541/541 [==============================] - 101s 187ms/step - loss: 0.0546 - accuracy: 0.7026 - val_loss: 0.0520 - val_accuracy: 0.7119
Epoch 3/15
541/541 [==============================] - ETA: 0s - loss: 0.0525 - accuracy: 0.7055
Epoch 00003: val_accuracy improved from 0.71191 to 0.71491, saving model to best_model_text.h5
541/541 [==============================] - 102s 188ms/step - loss: 0.0525 - accuracy: 0.7055 - val_loss: 0.0511 - val_accuracy: 0.7149
Epoch 4/15
541/5

#3)Use LSTM to classify noist text based on words split from noisy text:

In [ ]:
from __future__ import unicode_literals, print_function, division
from io import open
import unicodedata
import string
import re
import random
import math

import torch
import torch.nn as nn
from torch import optim
import torch.nn.functional as F

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

In [ ]:
train

id  ...                               noisyTextDescription
0      36274  ...  Femella Women Ankle-Length Grey AQ-S800WD-1EVD...
1      15129  ...              Converse Unisex Casual Skirts Slipper
2      58976  ...                          Velia Women Acetone Kurta
3      32922  ...                  Enroute Men Leather Brown Sandals
4      29561  ...              Aneri Exclusive Anu Pink Inspirartion
...      ...  ...                                                ...
21622   5435  ...                    SHE-3802D-1ADR Men Black Briefs
21623  27880  ...                 Adipower & Taylor Men Brown Quince
21624  38385  ...                             Rocia Women R348 Flats
21625  34853  ...                  Fastrack Men Black Amethyst Watch
21626  41407  ...             Arrow Ghicha Men White Slim Surf Shirt

[21627 rows x 7 columns]

In [ ]:
train['combine_text']=train["noisyTextDescription"].astype(str)+' '+train["gender"].astype(str)+' '+train['baseColour'].astype(str)+' '+train['season'].astype(str)+' '+train['usage'].astype(str)
#train['combine_text']=train['noisyTextDescription'].astype(str)
#train['combine_text']=train["gender"].astype(str)+' '+train['baseColour'].astype(str)+' '+train['season'].astype(str)+' '+train['usage'].astype(str)
train['combine_text'][:10]

0    Femella Women Ankle-Length Grey AQ-S800WD-1EVD...
1    Converse Unisex Casual Skirts Slipper Unisex G...
2    Velia Women Acetone Kurta Women Red Summer Ethnic
3    Enroute Men Leather Brown Sandals Men Brown Su...
4    Aneri Exclusive Anu Pink Inspirartion Women Pi...
5    Envirosax Women Multi Corp T2N150 Women Multi ...
6    Playboy Begining Tights Skt Men Black Summer C...
7    Femella Women White ComfortSoft Shirt Women Wh...
8    Avirate Women Blue Chanderi Dupatta Women Blue...
9    Lotto Women Virtue CSM LTP-1283D-2ADF Women Re...
Name: combine_text, dtype: object

In [ ]:
import numpy as np
list1 = [10, 20, 10, 30, 40, 40] 
print("the unique values from 1st list is") 
np.unique(list1) 
  

the unique values from 1st list is


array([10, 20, 30, 40])

In [ ]:
import unicodedata
import string
import math

all_letters = string.ascii_letters + " .,;'"
n_letters = len(all_letters)

In [ ]:
def normalizeString(s):
    s = unicodeToAscii(s.lower().strip())
    s = re.sub(r"([.!?])", r" \1", s)
    s = re.sub(r"[^a-zA-Z]+", r" ", s)
    return s

In [ ]:
def unicodeToAscii(s):
    return ''.join(
        c for c in unicodedata.normalize('NFD', s)
        if unicodedata.category(c) != 'Mn'
        and c in all_letters
    )

print(unicodeToAscii('Ślusàrski'))


Slusarski


In [ ]:
normalizeString("sdfs.df")

'sdfs df'

In [ ]:
a=["df"]
a.append("r")
a

['df', 'r']

In [ ]:
train['combine_text']

0          Women Grey Summer Casual
1        Unisex Green Summer Casual
2           Women Red Summer Ethnic
3           Men Brown Summer Casual
4            Women Pink Fall Ethnic
                    ...            
21622       Men Black Summer Casual
21623       Men Brown Summer Casual
21624     Women Beige Winter Casual
21625       Men Steel Winter Casual
21626       Men White Summer Casual
Name: combine_text, Length: 21627, dtype: object

In [ ]:
import numpy as np
category_lines={}
for i in range(train.shape[0]):
  if train['category'][i] not in category_lines:
       category_lines[train['category'][i]]=[normalizeString(train['combine_text'][i])]
  else:
       category_lines[train['category'][i]].append(normalizeString(train['combine_text'][i]))
for i in category_lines:
   category_lines[i]=np.unique(category_lines[i]).tolist()

In [ ]:
import numpy as np
category_lines={}
for i in range(train.shape[0]):
  if train['category'][i] not in category_lines:
       category_lines[train['category'][i]]=[train['combine_text'][i]]
  else:
       category_lines[train['category'][i]].append(train['combine_text'][i])
for i in category_lines:
   category_lines[i]=np.unique(category_lines[i]).tolist()

In [ ]:
train['category']

0           Scarves
1        Flip Flops
2           Topwear
3            Sandal
4           Topwear
            ...    
21622     Innerwear
21623         Belts
21624         Shoes
21625       Watches
21626       Topwear
Name: category, Length: 21627, dtype: object

In [ ]:
len(category_lines)

27

In [ ]:
category_lines['Topwear']

["'N Accessory COLOR Main Kurta Women White Summer Ethnic",
 '(A504) Machine Women W-210-1CVDF Grey Tshirts Women Grey Fall Casual',
 '(AD164) Tree Kids Girls Saturday Pink Tops Women Pink Fall Casual',
 '(SH162) Emperro Men HS Blue Hail Men Navy Blue Fall Casual',
 '+3 OUTLET Women Printed Navy Blue Kurti Women Navy Blue Fall Ethnic',
 '-Logic Men Check 18+ Shirt Men Red Fall Casual',
 '/ Colors of Benetton MR310 dunhill Green Tshirts Men Green Fall Casual',
 "/Crayon Men's Purple Polo T-shirt Men Black Summer Sports",
 '007 United Men Printed Yellow Jhuti Men Yellow Fall Casual',
 '023 Natural Unisex Hybrid Jacket Men Black Summer Casual',
 "027 Men's Lion Logo White T-shirt Men White Summer Casual",
 '028 Lotus Grp Benetton Women Printed Black & White Top Women Black Summer Casual',
 '029 Men Check Array peach Men Blue Fall Formal',
 '034 Men Purple 06 T-shirt Men Purple Fall Casual',
 '036 BCPP01 Round Neck Blue T-shirt Men Blue Summer Casual',
 '0380 Gentle Printed Green Kurta Wom

In [ ]:
# Build the category_lines dictionary, a list of names per language
train_data = {}
validation_data = {}
all_categories = []


for category in category_lines:
    all_categories.append(category)
    lines = category_lines[category]
    random.shuffle(lines)
    train_data[category] = lines[0:int(math.floor(0.8*len(lines)))]
    validation_data[category] = lines[int(math.floor(0.8*len(lines)))+1:]

n_categories = len(all_categories)

In [ ]:
train_data['Bags']

['living Women Envelope Green Handbag Women Green Summer Casual',
 'American Tourister Unisex Buzz Red Backpack Unisex Red Winter Casual',
 'Baggit Fluctuation Trailway Taj Maker Handbag Women Brown Winter Casual',
 'United Colors of Benetton Women Solid Red Swoosh Women Red Summer Casual',
 'Wildcraft Glorious Grey & Black Backpack Unisex Grey Winter Casual',
 'Baggit Women Torment Handbag Women Brown Summer Casual',
 'Wildcraft Unisex Red Backpack Unisex Red Winter Casual',
 'Envirosax Women Balti Blue CT Women Navy Blue Summer Casual',
 'INTENSE Women Purple Bag Women Purple Summer Casual',
 'Wildcraft Unisex Red Fenil Grey Rucksack Unisex Red Winter Casual',
 'Murcia Alvord Mauve Handbag Women Mauve Summer Casual',
 'Baggit Women Galaxy Exel Navy Blue Handbag Women Navy Blue Winter Casual',
 'Rocia Women Was Clutch Women Green Summer Casual',
 'Alphaballer Taupe Black & Orange Backpack Men Black Winter Casual',
 'Nyk Women Flaunted Vespa Handbag Women Black Summer Casual',
 'Fastra

In [ ]:
all_letters

"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ .,;'"

In [ ]:
category_lines

{'Accessories': ['Lino Perros Men Formal Black Accessory LCESCBRA Set',
  'Arrow Men Formal Purple Tie+Cufflink+Pocket square Mir Combo Pack',
  'Turtle Fold Purple Accessory Gift Set',
  'ADIDAS Phantom Black Sipper',
  'ToniQ Women Pink White-Silver Accessories',
  'Lino Sole Men Formal Blue Umbro Accessory Gift LTP-1233D-1ADF',
  'United Colors Intermeshed Benetton Women Solid Blue Cinder',
  'Vital Gear Unisex Black Messenger Alpinist',
  'Glitterati Unisex Black Seven Pouch',
  'Turtle Tempting Formal Maroon Yorker Panda Set',
  'Nba Unisex 039 Rosetti Bag',
  'Sahara Men Fraiche Frank Tie+Cufflink+Pocket square - Combo Chhota',
  'Lino Perros Men Formal Valaya Latino Gift Set',
  'Turtle Men U Accessory Gift Ftbll/Socc',
  'Belkin Unisex Rugged Messenger Bag',
  'Reid & Taylor Men Formal Red 836 square - Vini Pack',
  'Lino Freya Blug Formal Everlasting Accessory Gift Set',
  'Cabarelli Men Black G-SHOCK AE-1100W-1AVDF(D085) Set',
  'AM4405 Vivid Never Turquiose Women Caps Olive 

In [ ]:
print(category_lines['Innerwear'][:5])

['Amante Women Taupe Bikini Briefs PGSI01', 'Enamor Red Bra', 'Lovable Women Spice Blue EF-305D-1AVDF', 'Home Women White Bikini webbing PGSI01', 'Bike Men Mr.Silly Blue Trevor NH679SL02A']


In [ ]:
n_categories

27

In [ ]:
n_letters

57

In [ ]:
all_letters

"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ .,;'"

In [ ]:
import torch

# Find letter index from all_letters, e.g. "a" = 0
def letterToIndex(letter):
    return all_letters.find(letter)

# Just for demonstration, turn a letter into a <1 x n_letters> Tensor
def letterToTensor(letter):
    tensor = torch.zeros(1, n_letters)
    tensor[0][letterToIndex(letter)] = 1
    return tensor

# Turn a line into a <line_length x 1 x n_letters>,
# or an array of one-hot letter vectors
def lineToTensor(line):
    tensor = torch.zeros(len(line), 1, n_letters)
    for li, letter in enumerate(line):
        tensor[li][0][letterToIndex(letter)] = 1
    return tensor

def sentenceToTensor(sentence):
    tensor = torch.zeros(len(sentence), 1, n_letters)
    for li, letter in enumerate(sentence):
        tensor[li][0][letterToIndex(letter)] = 1
    return tensor
         

#print(letterToTensor('J')[0])

#print(lineToTensor('Jones'))

print(sentenceToTensor('Joes Harr'))

tensor([[[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
          0., 0., 0., 0., 0., 0

In [ ]:
class RNN_LSTM(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(RNN_LSTM, self).__init__()

        self.hidden_size = hidden_size

        self.LSTM=nn.LSTMCell(input_size,hidden_size)
        self.i2o = nn.Linear(input_size + hidden_size, output_size)
        self.softmax = nn.LogSoftmax(dim=1)

    def forward(self, input, hidden, cell):
        hidden, cell = self.LSTM(input, (hidden,cell))
        combined = torch.cat((input, hidden), 1)
        output = self.i2o(combined)
        output = self.softmax(output)
        return output,hidden,cell

    def initHidden(self):
        return torch.zeros(1, self.hidden_size)

n_hidden = 128
rnn_lstm = RNN_LSTM(n_letters, n_hidden, n_categories)  

In [ ]:
input = sentenceToTensor('Annetta Harry')
hidden = torch.zeros(1, n_hidden)
cell=torch.zeros(1, n_hidden)
output,hidden,cell= rnn_lstm(input[0], hidden,cell)
print(output)

tensor([[-3.3341, -3.2704, -3.2007, -3.1972, -3.3705, -3.3515, -3.1855, -3.3367,
         -3.2297, -3.3455, -3.3872, -3.2706, -3.2763, -3.2179, -3.3376, -3.3686,
         -3.4011, -3.2348, -3.3265, -3.2260, -3.2783, -3.3777, -3.2142, -3.2770,
         -3.2771, -3.3856, -3.3700]], grad_fn=<LogSoftmaxBackward>)


In [ ]:
def train_LSTM(category_tensor, line_tensor):
    hidden = rnn_lstm.initHidden()
    cell= rnn_lstm.initHidden()
    # reset gradient
    rnn_lstm.zero_grad()
    
    for i in range(line_tensor.size()[0]):
        output,hidden,cell = rnn_lstm(line_tensor[i], hidden, cell)
    loss = criterion_lstm(output, category_tensor)

    # compute gradient by backpropagation
    loss.backward()

    # update parameters
    optimizer_lstm.step()

    return output, loss.item()

In [ ]:
def categoryFromOutput(output):
    top_n, top_i = output.topk(1)
    category_i = top_i[0].item()
    return all_categories[category_i], category_i

print(categoryFromOutput(output))

('Shoes', 6)


In [ ]:
all_categories

['Scarves',
 'Flip Flops',
 'Topwear',
 'Sandal',
 'Bags',
 'Socks',
 'Shoes',
 'Watches',
 'Dress',
 'Headwear',
 'Jewellery',
 'Bottomwear',
 'Innerwear',
 'Wallets',
 'Belts',
 'Saree',
 'Nails',
 'Loungewear and Nightwear',
 'Lips',
 'Eyewear',
 'Makeup',
 'Ties',
 'Fragrance',
 'Cufflinks',
 'Free Gifts',
 'Apparel Set',
 'Accessories']

In [ ]:
import random

def randomChoice(l):
    return l[random.randint(0, len(l) - 1)]

def randomTrainingExample():
    category = randomChoice(all_categories)
    line = randomChoice(train_data[category])
    category_tensor = torch.tensor([all_categories.index(category)], dtype=torch.long)
    line_tensor = sentenceToTensor(line)
    return category, line, category_tensor, line_tensor

for i in range(10):
    category, line, category_tensor, line_tensor = randomTrainingExample()
    print('category =', category, '/ line =', line)

category = Socks / line = ADIDAS Unisex White Frujuvenate of 3 Socks
category = Belts / line = Wrangler Women Brown Legging
category = Fragrance / line = Valentino Perfumes Women Prima See Nehru Perfume
category = Lips / line = Colorbar Take Me As I 1/2 Mischievous N6015sm01 Mettalic 008
category = Shoes / line = F Sports Calcetto Navy Tunic Col Shoes
category = Accessories / line = Tenkay Men 129 & White DC Scholl
category = Innerwear / line = Peri Marches Women Black Bra
category = Watches / line = Titan Sphere Black Watch
category = Saree / line = FNF Multi Coloured 3HD Sole
category = Shoes / line = Bottle Men's Tarun Lowell Navy Silver Red HDD-600C-2AVDF(D057)


In [ ]:
criterion = nn.NLLLoss()

In [ ]:
import time
import math
from torch.autograd import Variable

criterion_lstm = nn.NLLLoss()
rnn_lstm = RNN_LSTM(n_letters, n_hidden, n_categories)  
optimizer_lstm = torch.optim.Adam(rnn_lstm.parameters())

n_iters = 80000
print_every = 5000
plot_every = 1000

# Keep track of losses for plotting
train_loss_lstm = 0
all_train_losses_lstm = []
all_validation_losses_lstm = []
all_validation_losses2_lstm = []

def timeSince(since):
    now = time.time()
    s = now - since
    m = math.floor(s / 60)
    s -= m * 60
    return '%dm %ds' % (m, s)

# Just return an output given a line
def evaluate_lstm(line_tensor):
    hidden = rnn_lstm.initHidden()
    cell= rnn_lstm.initHidden()
    for i in range(line_tensor.size()[0]):
        output,hidden,cell = rnn_lstm(line_tensor[i], hidden,cell)
    return output

def eval_dataset_lstm(dataset):
    loss = 0
    n_instances = 0
    confusion = torch.zeros(n_categories, n_categories)
    for category in all_categories:
        category_tensor = Variable(torch.LongTensor([all_categories.index(category)]))
        n_instances += len(dataset[category])
        for line in dataset[category]:
            line_tensor = Variable(sentenceToTensor(line))
            output = evaluate_lstm(line_tensor)
            loss += criterion_lstm(output, category_tensor)
            guess, guess_i = categoryFromOutput(output)
            category_i = all_categories.index(category)
            confusion[category_i][guess_i] += 1

    # Normalize by dividing every row by its sum
    for i in range(n_categories):
        confusion[i] = confusion[i] / confusion[i].sum()

    return loss.item() / n_instances, confusion
  
print('\nIter \tTrain% \tTime \t\tTrain_loss_lstm \tExample')
start = time.time()
for iter in range(1, n_iters + 1):
    category, line, category_tensor, line_tensor = randomTrainingExample()
    output, loss = train_LSTM(category_tensor, line_tensor)
    train_loss_lstm += loss

    # Print iter number, train loss average, name and guess
    if iter % print_every == 0:
        guess, guess_i = categoryFromOutput(output)
        correct = '✓' if guess == category else '✗ (%s)' % category
        print('%d \t%d%% \t(%s) \t%.4f \t\t%s / %s %s' % (iter, iter / n_iters * 100, timeSince(start), train_loss_lstm / plot_every, line, guess, correct))

    # Add current train loss average to list of losses
    if iter % plot_every == 0:
        all_train_losses_lstm.append(train_loss_lstm / plot_every)
        train_loss_lstm = 0
        
    # Compute loss based on validation data
    if iter % plot_every == 0:
        average_validation_loss_lstm, _ = eval_dataset_lstm(validation_data)

        # save model with best validation loss
        if len(all_validation_losses_lstm) == 0 or average_validation_loss_lstm < min(all_validation_losses_lstm):
            torch.save(rnn_lstm, 'char_rnn_lstm_classification_model.pt')
        all_validation_losses_lstm.append(average_validation_loss_lstm)



Iter 	Train% 	Time 		Train_loss_lstm 	Example
5000 	6% 	(4m 3s) 	1.4307 		Lakme HB65 Matte Satin Ruby Lipstick 42 / Lips ✓
10000 	12% 	(8m 44s) 	1.0454 		Femella Women Riveted Scarf / Scarves ✓
15000 	18% 	(13m 20s) 	0.9309 		Pieces Elements Mauve Ethnic Belt / Belts ✓
20000 	25% 	(17m 57s) 	0.8244 		Colorbar Turquoise Enchanting FC1019BWN Eye Shadow 007 / Makeup ✓
25000 	31% 	(22m 33s) 	0.7670 		Prem Absolute White Intense Golden Medium Wet & Dry Compact Sepia / Makeup ✓
30000 	37% 	(27m 7s) 	0.6909 		FNF Orange Printed Sari / Saree ✓
35000 	43% 	(31m 37s) 	0.5666 		Rasasi 365 AP1103 Flame Deo / Free Gifts ✓
40000 	50% 	(35m 59s) 	0.5208 		Nike Men Air Max+ 2011 SHE-4022D-4ADR Sports Shoes / Shoes ✓
45000 	56% 	(40m 15s) 	0.5519 		Rreverie Women Silver Earrings / Jewellery ✓
50000 	62% 	(44m 27s) 	0.5429 		Lakme True Petals Classics ED240 Red Nail Polish Years / Nails ✓
55000 	68% 	(48m 36s) 	0.4598 		ALayna Embroidary Printed White Kurta / Topwear ✓
60000 	75% 	(52m 57s) 	0.3930 		F

In [ ]:
# evaluate on test set and display confusion matrix
rnn_lstm = torch.load('char_rnn_lstm_classification_model.pt')
average_test_loss_lstm, confusion_lstm = eval_dataset_lstm(validation_data)
print('Average test loss: ' + str(average_test_loss_lstm))

Average test loss: 0.9977324065180486


In [ ]:
validation_data.keys()

dict_keys(['Scarves', 'Flip Flops', 'Topwear', 'Sandal', 'Bags', 'Socks', 'Shoes', 'Watches', 'Dress', 'Headwear', 'Jewellery', 'Bottomwear', 'Innerwear', 'Wallets', 'Belts', 'Saree', 'Nails', 'Loungewear and Nightwear', 'Lips', 'Eyewear', 'Makeup', 'Ties', 'Fragrance', 'Cufflinks', 'Free Gifts', 'Apparel Set', 'Accessories'])

In [ ]:
def predict(input_line, n_predictions=3):
    #print('\n> %s' % input_line)
    with torch.no_grad():
        output = evaluate_lstm(sentenceToTensor(input_line))

        # Get top N categories
        topv, topi = output.topk(n_predictions, 1, True)
        predictions = []

        for i in range(n_predictions):
            value = topv[0][i].item()
            category_index = topi[0][i].item()
            #print(all_categories[category_index])
            predictions.append(all_categories[category_index])
        return(predictions)
predict('women white winter casual')


['Saree', 'Sandal', 'Bottomwear']

In [ ]:
test['combine_text']=test['noisyTextDescription'].astype(str)
#test['combine_text']=test["gender"].astype(str)+' '+test['baseColour'].astype(str)+' '+test['season'].astype(str)+' '+test['usage'].astype(str)
test['combine_text']

0                        Chromozome Men Black Fashion Vest
1                         Elle Women Green Color Clash Top
2                     Baggit Women Chotu Mayur Black Palms
3        Greensboro Colors Of Ap Men Latnam Black Casua...
4                          Fish White Smooth daddy (SH100)
                               ...                        
21623                  Park ClubLife 602 Helium Sunglasses
21624                Vero Moda Women Black Rey Shift Dress
21625                     Nike Men Hell With White Tshirts
21626         Lakme Holi to Five Day Apricot Nectar Lip K9
21627                    Percorso Unisex Red Grit Backpack
Name: combine_text, Length: 21628, dtype: object

In [ ]:
train.shape

(21627, 8)

In [ ]:
pred=[]
accu=0
for line in train['combine_text']:
  pred.extend(predict(line,n_predictions=1))
for predi,output in zip(pred,train['category']):
  if predi==output:
      accu+=1
print(accu/train.shape[0])
pred

0.7800897026864567


['Scarves',
 'Sandal',
 'Topwear',
 'Sandal',
 'Wallets',
 'Saree',
 'Belts',
 'Topwear',
 'Topwear',
 'Watches',
 'Scarves',
 'Bags',
 'Topwear',
 'Watches',
 'Topwear',
 'Topwear',
 'Watches',
 'Topwear',
 'Sandal',
 'Ties',
 'Sandal',
 'Topwear',
 'Shoes',
 'Headwear',
 'Topwear',
 'Jewellery',
 'Bottomwear',
 'Topwear',
 'Headwear',
 'Bags',
 'Bags',
 'Topwear',
 'Bags',
 'Innerwear',
 'Belts',
 'Wallets',
 'Fragrance',
 'Sandal',
 'Saree',
 'Topwear',
 'Watches',
 'Topwear',
 'Scarves',
 'Jewellery',
 'Shoes',
 'Jewellery',
 'Wallets',
 'Shoes',
 'Dress',
 'Sandal',
 'Shoes',
 'Innerwear',
 'Topwear',
 'Topwear',
 'Eyewear',
 'Nails',
 'Topwear',
 'Sandal',
 'Topwear',
 'Innerwear',
 'Topwear',
 'Topwear',
 'Shoes',
 'Topwear',
 'Topwear',
 'Socks',
 'Loungewear and Nightwear',
 'Topwear',
 'Shoes',
 'Innerwear',
 'Fragrance',
 'Shoes',
 'Flip Flops',
 'Topwear',
 'Topwear',
 'Lips',
 'Bags',
 'Sandal',
 'Topwear',
 'Shoes',
 'Topwear',
 'Topwear',
 'Shoes',
 'Nails',
 'Shoes',
 '

In [ ]:
pred=[]
for line in test['combine_text']:
  pred.extend(predict(line,n_predictions=1))

In [ ]:
predict=np.stack((test['id'],np.array(pred)),axis=1)
import csv

with open('submission.csv', mode='w') as csv_file:
    writer = csv.writer(csv_file)
    writer.writerow(['id','category'])
    writer.writerows(predict)

In [ ]:
predict.shape

(21628, 2)

In [ ]:
pwd

'/'

In [ ]:
cd /content/drive/MyDrive/Colab\ Notebooks/CS680\ Kaggle/

/content/drive/MyDrive/Colab Notebooks/CS680 Kaggle


#4)use attention to classify labels only:

In [ ]:
 import pandas as pd
train=pd.read_csv('train.csv',header=0)
print(train.columns)
train.head#21627

Index(['id', 'category', 'gender', 'baseColour', 'season', 'usage',
       'noisyTextDescription'],
      dtype='object')


<bound method NDFrame.head of           id  ...                               noisyTextDescription
0      36274  ...  Femella Women Ankle-Length Grey AQ-S800WD-1EVD...
1      15129  ...              Converse Unisex Casual Skirts Slipper
2      58976  ...                          Velia Women Acetone Kurta
3      32922  ...                  Enroute Men Leather Brown Sandals
4      29561  ...              Aneri Exclusive Anu Pink Inspirartion
...      ...  ...                                                ...
21622   5435  ...                    SHE-3802D-1ADR Men Black Briefs
21623  27880  ...                 Adipower & Taylor Men Brown Quince
21624  38385  ...                             Rocia Women R348 Flats
21625  34853  ...                  Fastrack Men Black Amethyst Watch
21626  41407  ...             Arrow Ghicha Men White Slim Surf Shirt

[21627 rows x 7 columns]>

In [ ]:
train['category']

0           Scarves
1        Flip Flops
2           Topwear
3            Sandal
4           Topwear
            ...    
21622     Innerwear
21623         Belts
21624         Shoes
21625       Watches
21626       Topwear
Name: category, Length: 21627, dtype: object

In [ ]:
test=pd.read_csv('test.csv',header=0)
print(test.columns)
test.head#21627

Index(['id', 'gender', 'baseColour', 'season', 'usage',
       'noisyTextDescription'],
      dtype='object')


<bound method NDFrame.head of           id  gender  ...   usage                               noisyTextDescription
0      26266     Men  ...  Casual                  Chromozome Men Black Fashion Vest
1      22134   Women  ...  Casual                   Elle Women Green Color Clash Top
2      28358   Women  ...  Casual               Baggit Women Chotu Mayur Black Palms
3      15554     Men  ...  Casual  Greensboro Colors Of Ap Men Latnam Black Casua...
4      53408   Women  ...  Casual                    Fish White Smooth daddy (SH100)
...      ...     ...  ...     ...                                                ...
21623  39737     Men  ...  Casual                Park ClubLife 602 Helium Sunglasses
21624  57477   Women  ...  Casual              Vero Moda Women Black Rey Shift Dress
21625  22312     Men  ...  Sports                   Nike Men Hell With White Tshirts
21626  54105   Women  ...  Casual       Lakme Holi to Five Day Apricot Nectar Lip K9
21627  14080  Unisex  ...  Casual  

Many test noisy text contains words that are not included in training set noisy text, we need to remove them:

In [ ]:
train_corpus=[]
for line in train['noisyTextDescription']:
  lst=line.split(" ")
  train_corpus.extend(lst)
corpus=np.unique(train_corpus).tolist()

In [ ]:
test_noisy=[]
for line in test['noisyTextDescription']:
  lst=line.split(" ")
  new_lst=""
  for word in lst:
    if word in corpus:
      new_lst=new_lst+" "+ word
  test_noisy.append(new_lst)

In [ ]:
test['noisyTextDescription']=test_noisy


In [ ]:
test['noisyTextDescription']

0                        Chromozome Men Black Fashion Vest
1                         Elle Women Green Color Clash Top
2                     Baggit Women Chotu Mayur Black Palms
3         Greensboro Colors Of Ap Men Latnam Black Casu...
4                          Fish White Smooth daddy (SH100)
                               ...                        
21623                  Park ClubLife 602 Helium Sunglasses
21624                Vero Moda Women Black Rey Shift Dress
21625                     Nike Men Hell With White Tshirts
21626         Lakme Holi to Five Day Apricot Nectar Lip K9
21627                    Percorso Unisex Red Grit Backpack
Name: noisyTextDescription, Length: 21628, dtype: object

In [ ]:
# train['combine_text']=train["gender"].astype(str)+' '+train['baseColour'].astype(str)+' '+train['season'].astype(str)+' '+train['usage'].astype(str)+' '+train['noisyTextDescription'].astype(str)+"\t"+train['category'].astype(str)
# test['combine_text']=test["gender"].astype(str)+' '+test['baseColour'].astype(str)+' '+test['season'].astype(str)+' '+test['usage'].astype(str)+' '+test['noisyTextDescription'].astype(str)
train['combine_text']=train['noisyTextDescription']+"\t"+train['category']
test['combine_text']=test['noisyTextDescription']
train['combine_text'][:10]

0    Femella Women Ankle-Length Grey AQ-S800WD-1EVD...
1    Converse Unisex Casual Skirts Slipper\tFlip Flops
2                   Velia Women Acetone Kurta\tTopwear
3            Enroute Men Leather Brown Sandals\tSandal
4       Aneri Exclusive Anu Pink Inspirartion\tTopwear
5              Envirosax Women Multi Corp T2N150\tBags
6                   Playboy Begining Tights Skt\tSocks
7       Femella Women White ComfortSoft Shirt\tTopwear
8         Avirate Women Blue Chanderi Dupatta\tTopwear
9         Lotto Women Virtue CSM LTP-1283D-2ADF\tShoes
Name: combine_text, dtype: object

In [ ]:
with open('noisy-category.txt', 'w') as f:
    for item in train['combine_text']:
        f.write("%s\n" % item)

In [ ]:
from __future__ import unicode_literals, print_function, division
from io import open
import unicodedata
import string
import re
import random
import math

import torch
import torch.nn as nn
from torch import optim
import torch.nn.functional as F

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

In [ ]:
SOS_token = 0
EOS_token = 1


class Lang:
    def __init__(self, name):
        self.name = name
        self.word2index = {}
        self.word2count = {}
        self.index2word = {0: "SOS", 1: "EOS"}
        self.n_words = 2  # Count SOS and EOS

    def addSentence(self, sentence):
        for word in sentence.split(' '):
            self.addWord(word)

    def addWord(self, word):
        if word not in self.word2index:
            self.word2index[word] = self.n_words
            self.word2count[word] = 1
            self.index2word[self.n_words] = word
            self.n_words += 1
        else:
            self.word2count[word] += 1

In [ ]:
# Turn a Unicode string to plain ASCII, thanks to
# http://stackoverflow.com/a/518232/2809427
def unicodeToAscii(s):
    return ''.join(
        c for c in unicodedata.normalize('NFD', s)
        if unicodedata.category(c) != 'Mn'
    )

# Lowercase, trim, and remove non-letter characters


def normalizeString(s):
    s = unicodeToAscii(s.lower().strip())
    s = re.sub(r"([.!?])", r" \1", s)
    s = re.sub(r"[^a-zA-Z.!?]+", r" ", s)
    return s

In [ ]:
def readLangs(lang1, lang2, reverse=False):
    print("Reading lines...")

    # Read the file and split into lines
    lines = open('%s-%s.txt' % (lang1, lang2), encoding='utf-8').\
        read().strip().split('\n')

    # Split every line into pairs and normalize
    pairs = [[s for s in l.split('\t')] for l in lines]

    # Reverse pairs, make Lang instances
    if reverse:
        pairs = [list(reversed(p)) for p in pairs]
        input_lang = Lang(lang2)
        output_lang = Lang(lang1)
    else:
        input_lang = Lang(lang1)
        output_lang = Lang(lang2)

    return input_lang, output_lang, pairs

In [ ]:
MAX_LENGTH = 100

eng_prefixes = (
    "i am ", "i m ",
    "he is", "he s ",
    "she is", "she s",
    "you are", "you re ",
    "we are", "we re ",
    "they are", "they re "
)


def filterPair(p):
    return len(p[0].split(' ')) < MAX_LENGTH and \
        len(p[1].split(' ')) < MAX_LENGTH and \
        p[1].startswith(eng_prefixes)


def filterPairs(pairs):
    return [pair for pair in pairs if filterPair(pair)]

In [ ]:
input_lang, output_lang, pairs =readLangs("noisy", "category")

Reading lines...


In [ ]:
def prepareData(lang1, lang2, reverse=False):
    input_lang, output_lang, pairs = readLangs(lang1, lang2, reverse)
    print("Read %s sentence pairs" % len(pairs))
    #pairs = filterPairs(pairs)
    print("Trimmed to %s sentence pairs" % len(pairs))
    print("Counting words...")
    for pair in pairs:
        input_lang.addSentence(pair[0])
        output_lang.addSentence(pair[1])
    print("Counted words:")
    print(input_lang.name, input_lang.n_words)
    print(output_lang.name, output_lang.n_words)
    return input_lang, output_lang, pairs


input_lang, output_lang, pairs = prepareData('noisy', 'category', False)
random.shuffle(pairs)
#train_data = pairs[0:int(math.floor(0.9*len(pairs)))]
#test_data = pairs[int(math.floor(0.9*len(pairs)))+1:]
train_data = pairs[0:int(math.floor(0.8*len(pairs)))]
validation_data = pairs[int(math.floor(0.8*len(pairs)))+1:]
#test_data = pairs[int(math.floor(0.9*len(pairs)))+1:]

print(random.choice(pairs))

Reading lines...
Read 21627 sentence pairs
Trimmed to 21627 sentence pairs
Counting words...
Counted words:
noisy 9286
category 34
['Puma Men Black T-shirt', 'Topwear']


In [ ]:
class EncoderRNN(nn.Module):
    def __init__(self, input_size, hidden_size):
        super(EncoderRNN, self).__init__()
        self.hidden_size = hidden_size

        self.embedding = nn.Embedding(input_size, hidden_size)
        self.gru = nn.GRU(hidden_size, hidden_size)

    def forward(self, input, hidden):
        embedded = self.embedding(input).view(1, 1, -1)
        output = embedded
        output, hidden = self.gru(output, hidden)
        return output, hidden

    def initHidden(self):
        return torch.zeros(1, 1, self.hidden_size, device=device)

In [ ]:
class DecoderRNN(nn.Module):
    def __init__(self, hidden_size, output_size):
        super(DecoderRNN, self).__init__()
        self.hidden_size = hidden_size

        self.embedding = nn.Embedding(output_size, hidden_size)
        self.gru = nn.GRU(hidden_size, hidden_size)
        self.out = nn.Linear(hidden_size, output_size)
        self.softmax = nn.LogSoftmax(dim=1)

    def forward(self, input, hidden):
        output = self.embedding(input).view(1, 1, -1)
        output = F.relu(output)
        output, hidden = self.gru(output, hidden)
        output = self.softmax(self.out(output[0]))
        return output, hidden

    def initHidden(self):
        return torch.zeros(1, 1, self.hidden_size, device=device)

In [ ]:
class AttnDecoderRNN(nn.Module):
    def __init__(self, hidden_size, output_size, dropout_p=0.1, max_length=MAX_LENGTH):
        super(AttnDecoderRNN, self).__init__()
        self.hidden_size = hidden_size
        self.output_size = output_size
        self.dropout_p = dropout_p
        self.max_length = max_length

        self.embedding = nn.Embedding(self.output_size, self.hidden_size)
        self.attn = nn.Linear(self.hidden_size * 2, self.max_length)
        self.attn_combine = nn.Linear(self.hidden_size * 2, self.hidden_size)
        self.dropout = nn.Dropout(self.dropout_p)
        self.gru = nn.GRU(self.hidden_size, self.hidden_size)
        self.out = nn.Linear(self.hidden_size, self.output_size)

    def forward(self, input, hidden, encoder_outputs):
        embedded = self.embedding(input).view(1, 1, -1)
        embedded = self.dropout(embedded)

        attn_weights = F.softmax(
            self.attn(torch.cat((embedded[0], hidden[0]), 1)), dim=1)
        attn_applied = torch.bmm(attn_weights.unsqueeze(0),
                                 encoder_outputs.unsqueeze(0))

        output = torch.cat((embedded[0], attn_applied[0]), 1)
        output = self.attn_combine(output).unsqueeze(0)

        output = F.relu(output)
        output, hidden = self.gru(output, hidden)

        output = F.log_softmax(self.out(output[0]), dim=1)
        return output, hidden, attn_weights

    def initHidden(self):
        return torch.zeros(1, 1, self.hidden_size, device=device)

In [ ]:
def indexesFromSentence(lang, sentence):
    return [lang.word2index[word] for word in sentence.split(' ')]


def tensorFromSentence(lang, sentence):
    indexes = indexesFromSentence(lang, sentence)
    indexes.append(EOS_token)
    return torch.tensor(indexes, dtype=torch.long, device=device).view(-1, 1)


def tensorsFromPair(pair):
    input_tensor = tensorFromSentence(input_lang, pair[0])
    target_tensor = tensorFromSentence(output_lang, pair[1])
    return (input_tensor, target_tensor)

In [ ]:
teacher_forcing_ratio = 0.5


def train(input_tensor, target_tensor, encoder, decoder, encoder_optimizer, decoder_optimizer, criterion, max_length=MAX_LENGTH):
    encoder_hidden = encoder.initHidden()

    encoder_optimizer.zero_grad()
    decoder_optimizer.zero_grad()

    input_length = input_tensor.size(0)
    target_length = target_tensor.size(0)

    encoder_outputs = torch.zeros(max_length, encoder.hidden_size, device=device)

    loss = 0

    for ei in range(input_length):
        encoder_output, encoder_hidden = encoder(
            input_tensor[ei], encoder_hidden)
        encoder_outputs[ei] = encoder_output[0, 0]

    decoder_input = torch.tensor([[SOS_token]], device=device)

    decoder_hidden = encoder_hidden

    use_teacher_forcing = True if random.random() < teacher_forcing_ratio else False

    if use_teacher_forcing:
        # Teacher forcing: Feed the target as the next input
        for di in range(target_length):
            decoder_output, decoder_hidden, decoder_attention = decoder(
                decoder_input, decoder_hidden, encoder_outputs)
            loss += criterion(decoder_output, target_tensor[di])
            decoder_input = target_tensor[di]  # Teacher forcing

    else:
        # Without teacher forcing: use its own predictions as the next input
        for di in range(target_length):
            decoder_output, decoder_hidden, decoder_attention = decoder(
                decoder_input, decoder_hidden, encoder_outputs)
            topv, topi = decoder_output.topk(1)
            decoder_input = topi.squeeze().detach()  # detach from history as input

            loss += criterion(decoder_output, target_tensor[di])
            if decoder_input.item() == EOS_token:
                break

    loss.backward()

    encoder_optimizer.step()
    decoder_optimizer.step()

    return loss.item() / target_length

In [ ]:
import time
import math

def timeSince(since):
    now = time.time()
    s = now - since
    m = math.floor(s / 60)
    s -= m * 60
    return '%dm %ds' % (m, s)  
  
def eval(input_tensor, target_tensor, encoder, decoder, criterion, max_length=MAX_LENGTH):
    encoder_hidden = encoder.initHidden()

    input_length = input_tensor.size(0)
    target_length = target_tensor.size(0)

    encoder_outputs = torch.zeros(max_length, encoder.hidden_size, device=device)

    loss = 0

    for ei in range(input_length):
        encoder_output, encoder_hidden = encoder(
            input_tensor[ei], encoder_hidden)
        encoder_outputs[ei] = encoder_output[0, 0]

    decoder_input = torch.tensor([[SOS_token]], device=device)

    decoder_hidden = encoder_hidden

    # Without teacher forcing: use its own predictions as the next input
    for di in range(target_length):
        decoder_output, decoder_hidden, decoder_attention = decoder(
            decoder_input, decoder_hidden, encoder_outputs)
        topv, topi = decoder_output.topk(1)
        decoder_input = topi.squeeze().detach()  # detach from history as input

        loss += criterion(decoder_output, target_tensor[di])
        if decoder_input.item() == EOS_token:
            break

    return loss.item() / target_length

def eval_dataset(dataset, encoder, decoder, criterion):
    loss = 0
    for pair in dataset:
        pair_tensor = tensorsFromPair(pair)
        input_tensor = pair_tensor[0]
        target_tensor = pair_tensor[1]
        loss += eval(input_tensor, target_tensor, encoder, decoder, criterion)
    return loss / len(dataset)


In [ ]:
def trainIters(encoder, decoder, n_iters, print_every=2000, plot_every=1000):
    start = time.time()
    all_train_losses = []
    all_validation_losses = []
    train_loss = 0  # Reset every plot_every

    encoder_optimizer = torch.optim.Adam(encoder.parameters())
    decoder_optimizer = torch.optim.Adam(decoder.parameters())
    training_pairs = [tensorsFromPair(random.choice(train_data))
                      for i in range(n_iters)]
    criterion = nn.NLLLoss()

    print('\nIter \tTrain% \tTime \t\tTrain_loss')
    for iter in range(1, n_iters + 1):
        training_pair = training_pairs[iter - 1]
        input_tensor = training_pair[0]
        target_tensor = training_pair[1]

        loss = train(input_tensor, target_tensor, encoder,
                     decoder, encoder_optimizer, decoder_optimizer, criterion)
        train_loss += loss

        if iter % print_every == 0:
            print('%d \t%d%% \t(%s) \t%.4f' % (iter, iter / n_iters * 100, timeSince(start), train_loss / plot_every))

        if iter % plot_every == 0:
            all_train_losses.append(train_loss / plot_every)
            train_loss = 0
            
        # Compute loss based on validation data
        if iter % plot_every == 0:
            average_validation_loss = eval_dataset(validation_data, encoder, decoder, criterion)

            # save model with best validation loss
            if len(all_validation_losses) == 0 or average_validation_loss < min(all_validation_losses):
                torch.save(encoder, 'seq2seq_translation_encoder_model.pt')
                torch.save(attn_decoder, 'seq2seq_translation_attn_decoder_model.pt')
            all_validation_losses.append(average_validation_loss)

            #validation_loss = 0
            #for validation_pair in validation_data:
            #    validation_pair_tensor = tensorsFromPair(validation_pair)
            #    input_tensor = validation_pair_tensor[0]
            #    target_tensor = validation_pair_tensor[1]
            #    validation_loss += evaluate(input_tensor, target_tensor, encoder, decoder, criterion)
            #all_validation_losses.append(validation_loss / len(validation_data))
    return (all_train_losses,all_validation_losses)
            
    

In [ ]:
def evaluate(encoder, decoder, sentence, max_length=MAX_LENGTH):
    with torch.no_grad():
        input_tensor = tensorFromSentence(input_lang, sentence)
        input_length = input_tensor.size()[0]
        encoder_hidden = encoder.initHidden()

        encoder_outputs = torch.zeros(max_length, encoder.hidden_size, device=device)

        for ei in range(input_length):
            encoder_output, encoder_hidden = encoder(input_tensor[ei],
                                                     encoder_hidden)
            encoder_outputs[ei] += encoder_output[0, 0]

        decoder_input = torch.tensor([[SOS_token]], device=device)  # SOS

        decoder_hidden = encoder_hidden

        decoded_words = []
        decoder_attentions = torch.zeros(max_length, max_length)

        for di in range(max_length):
            decoder_output, decoder_hidden, decoder_attention = decoder(
                decoder_input, decoder_hidden, encoder_outputs)
            decoder_attentions[di] = decoder_attention.data
            topv, topi = decoder_output.data.topk(1)
            if topi.item() == EOS_token:
                decoded_words.append('<EOS>')
                break
            else:
                decoded_words.append(output_lang.index2word[topi.item()])

            decoder_input = topi.squeeze().detach()

        return decoded_words, decoder_attentions[:di + 1]

In [ ]:
def evaluateRandomly(encoder, decoder, n=10):
    for i in range(n):
        pair = random.choice(pairs)
        print('>', pair[0])
        print('=', pair[1])
        output_words, attentions = evaluate(encoder, decoder, pair[0])
        output_sentence = ' '.join(output_words)
        print('<', output_sentence)
        print('')

In [ ]:
hidden_size = 256
encoder = EncoderRNN(input_lang.n_words, hidden_size).to(device)
attn_decoder = AttnDecoderRNN(hidden_size, output_lang.n_words, dropout_p=0.1).to(device)
all_train_losses,all_validation_losses=trainIters(encoder, attn_decoder, 50000)


Iter 	Train% 	Time 		Train_loss
2000 	4% 	(1m 8s) 	0.4255
4000 	8% 	(2m 28s) 	0.3240
6000 	12% 	(3m 48s) 	0.2961
8000 	16% 	(5m 9s) 	0.2326
10000 	20% 	(6m 29s) 	0.1822
12000 	24% 	(7m 50s) 	0.1715
14000 	28% 	(9m 10s) 	0.1458
16000 	32% 	(10m 30s) 	0.1644
18000 	36% 	(11m 50s) 	0.1510
20000 	40% 	(13m 11s) 	0.1599
22000 	44% 	(14m 31s) 	0.1252
24000 	48% 	(15m 51s) 	0.1274
26000 	52% 	(17m 11s) 	0.0887
28000 	56% 	(18m 31s) 	0.1138
30000 	60% 	(19m 51s) 	0.0919
32000 	64% 	(21m 11s) 	0.0876
34000 	68% 	(22m 31s) 	0.0838
36000 	72% 	(23m 51s) 	0.0802
38000 	76% 	(25m 11s) 	0.0837
40000 	80% 	(26m 31s) 	0.0805
42000 	84% 	(27m 50s) 	0.0881
44000 	88% 	(29m 10s) 	0.0738
46000 	92% 	(30m 30s) 	0.0717
48000 	96% 	(31m 50s) 	0.0765
50000 	100% 	(33m 10s) 	0.0640


In [ ]:
# evaluate on test set and print average test loss
encoder = torch.load('seq2seq_translation_encoder_model.pt')
attn_decoder = torch.load('seq2seq_translation_attn_decoder_model.pt')
#average_test_loss = eval_dataset(train_data, encoder, attn_decoder, nn.NLLLoss())
#print('Average test loss: ' + str(average_test_loss))

AttributeError: ignored

In [ ]:
evaluateRandomly(encoder, attn_decoder)

> United Colors of Benetton Kids Girls Red Printed Top
= Topwear
< Topwear <EOS>

> Penny Unisex Beach Black Flip Flop
= Flip Flops
< Flip <EOS>

> Neckline Polo Men's Tube Yellow T-shirt
= Topwear
< Topwear <EOS>

> Seaweed Women Invigorate Flats
= Shoes
< Shoes <EOS>

> Red Tape Men Peep Corporate Light Brown Shoe
= Shoes
< Shoes <EOS>

> Jockey SP-OW Men Black Innerwear Vest SP11
= Innerwear
< Innerwear <EOS>

> Nike Men Challenger P161BK2F Jersey
= Topwear
< Topwear <EOS>

> New Hide EF-558D-7AVDF(ED438) Wowhol Wallet
= Wallets
< Wallets <EOS>

> Fastrack Unisex Sport Blue Sunglasses
= Eyewear
< Eyewear <EOS>

> Casio ENTICER Women 949 Re-incarnation 949 A254 LTP-1283D-2ADF
= Watches
< Bags <EOS>



In [ ]:
validation_data

[['Casio Enticer Notch Silver Multi-Dial Analogue Watch Dunhill', 'Watches'],
 ['Esprit Men Gran Orbus Black Steel Watches', 'Watches'],
 ['Nike Men With Stripe Avani Os', 'Topwear'],
 ['Photo Men Grey Multi-Function Shirt', 'Topwear'],
 ['Baggit Women Barrel Rosy Black Mobile Misha', 'Bags'],
 ['Van Heusen Field Black Sili Extreme Medium Belt', 'Belts'],
 ['Fabindia Strappy Brown Cotton Nutranite Dupatta', 'Topwear'],
 ['ADIDAS Originals Women Superstar 2 White Casual Keisha', 'Shoes'],
 ['660 Colors of Benetton Women Brown Jumpsuit', 'Dress'],
 ['Warior Megan Purple Fierce T-shirt', 'Topwear'],
 ['FOSSIL Men Red Tie', 'Ties'],
 ['Flair Authority Men Aerial Green Alicia', 'Topwear'],
 ["Watche Men's Rakett Black Shoe", 'Shoes'],
 ['ADIDAS Originals Men Solid Black Personalities', 'Topwear'],
 ['Nike Men Laertes Brown HA! Flops', 'Flip Flops'],
 ['French Villa Women Burgundy Top', 'Topwear'],
 ['Desman Men Brown Socks', 'Socks'],
 ['Proline Men Navy Striped Polo T-shirt', 'Topwear'],
 

In [ ]:
output_words, attentions = evaluate(
        encoder, attn_decoder, "Women Brown Summer Casual")
output_words

['Shoes', '<EOS>']

In [ ]:
import numpy as np
list_of_cat=np.unique(train['category'])
len(list_of_cat)

27

In [ ]:
accu=0
for i in train_data:
  cat=i[1]
  descript=i[0]
  print(cat)
  output_words, attentions = evaluate(encoder, attn_decoder, descript)
  #print(output_words[0])
  rough_pred=output_words[0]
  for j in list_of_cat:
    if  rough_pred.split(" ")[0]==j.split(" ")[0]:
        rough_pred=j
  print(rough_pred)
  if rough_pred==cat:
        accu+=1 
accu/len(train_data)

Streaming output truncated to the last 5000 lines.
Fragrance
Fragrance
Fragrance
Fragrance
Socks
Socks
Bags
Bags
Shoes
Shoes
Innerwear
Innerwear
Shoes
Shoes
Apparel Set
Apparel Set
Innerwear
Innerwear
Topwear
Topwear
Bottomwear
Bottomwear
Topwear
Topwear
Shoes
Shoes
Topwear
Topwear
Wallets
Wallets
Topwear
Topwear
Topwear
Topwear
Shoes
Shoes
Topwear
Topwear
Wallets
Topwear
Bags
Bags
Fragrance
Fragrance
Watches
Watches
Shoes
Shoes
Topwear
Topwear
Ties
Ties
Watches
Watches
Watches
Watches
Innerwear
Innerwear
Bags
Bags
Loungewear and Nightwear
Loungewear and Nightwear
Topwear
Topwear
Bags
Bags
Bags
Bags
Watches
Watches
Sandal
Sandal
Sandal
Sandal
Sandal
Topwear
Nails
Nails
Topwear
Topwear
Topwear
Topwear
Bottomwear
Bottomwear
Innerwear
Innerwear
Bags
Bags
Topwear
Topwear
Topwear
Topwear
Shoes
Shoes
Topwear
Topwear
Shoes
Shoes
Innerwear
Innerwear
Bottomwear
Bottomwear
Bottomwear
Bottomwear
Fragrance
Fragrance
Shoes
Shoes
Sandal
Sandal
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear


In [ ]:
accu/len(train_data)

0.9457834807236576

In [ ]:
accu=0
for i in validation_data:
  cat=i[1]
  descript=i[0]
  print(cat)
  output_words, attentions = evaluate(encoder, attn_decoder, descript)
  rough_pred=output_words[0]
  for j in list_of_cat:
    if  rough_pred.split(" ")[0]==j.split(" ")[0]:
        rough_pred=j
  print(rough_pred)
  if rough_pred==cat:
        accu+=1 
print(accu/len(validation_data))

Streaming output truncated to the last 5000 lines.
Topwear
Shoes
Shoes
Bottomwear
Bottomwear
Shoes
Shoes
Topwear
Topwear
Bags
Bags
Bags
Bags
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Shoes
Shoes
Watches
Topwear
Topwear
Topwear
Topwear
Topwear
Flip Flops
Topwear
Wallets
Jewellery
Shoes
Bags
Bags
Bags
Bags
Bags
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Innerwear
Innerwear
Flip Flops
Sandal
Shoes
Shoes
Topwear
Topwear
Bags
Bags
Bags
Bags
Bags
Bags
Topwear
Topwear
Shoes
Shoes
Watches
Watches
Dress
Dress
Jewellery
Jewellery
Bags
Watches
Jewellery
Jewellery
Topwear
Topwear
Belts
Belts
Jewellery
Saree
Topwear
Topwear
Shoes
Shoes
Shoes
Topwear
Topwear
Topwear
Bags
Bags
Watches
Watches
Topwear
Topwear
Wallets
Wallets
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Socks
Socks
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Bags
Bags
Topwear
Topwear
Topwear
Topwear
Bags
Top

In [ ]:
train_data[0]

['Women Black Summer Casual Numero Uno Women 65CC Black Shoe', 'Shoes']

In [ ]:
test=pd.read_csv('test.csv',header=0)
print(test.columns)
test.head#21627
test['noisyTextDescription']=test_noisy



Index(['id', 'gender', 'baseColour', 'season', 'usage',
       'noisyTextDescription'],
      dtype='object')


In [ ]:
for i in train_data:
  if "MTP-1291D-1A3VDF(A415)" in i[0]:
    print(i[0])

Men Black Summer Casual MTP-1291D-1A3VDF(A415) Men Brown And Black ES3069 Belt
Men Black Winter Formal Lino MTP-1291D-1A3VDF(A415) Men Formal Black Accessory Hadley Set


In [ ]:
def normalizeString(s):
    s = unicodeToAscii(s.strip())
    #s = re.sub(r"([.!?])", r" \1", s)
    s = re.sub(r"[&]+", r"", s)
    
    return s

In [ ]:
train['noisyTextDescription'][:3]

0    Femella Women Ankle-Length Grey AQ-S800WD-1EVD...
1                Converse Unisex Casual Skirts Slipper
2                            Velia Women Acetone Kurta
Name: noisyTextDescription, dtype: object

In [ ]:
test_noisy[0].strip()

'Chromozome Men Black Fashion Vest'

In [ ]:
pred_labels=[]
for i in test_noisy:
  #print(i)
  output_words, attentions = evaluate(encoder, attn_decoder,i.strip())
  #print(output_words[0])
  rough_pred=output_words[0]
  for j in list_of_cat:
    if  rough_pred.split(" ")[0]==j.split(" ")[0]:
        rough_pred=j
  print(rough_pred)
  pred_labels.append(rough_pred)  

Streaming output truncated to the last 5000 lines.
Shoes
Innerwear
Topwear
Watches
Topwear
Topwear
Innerwear
Loungewear and Nightwear
Topwear
Topwear
Bottomwear
Topwear
Shoes
Shoes
Bottomwear
Shoes
Headwear
Shoes
Topwear
Shoes
Cufflinks
Bags
Topwear
Topwear
Topwear
Shoes
Eyewear
Watches
Topwear
Bags
Topwear
Topwear
Topwear
Shoes
Shoes
Flip Flops
Bottomwear
Socks
Topwear
Bags
Topwear
Shoes
Shoes
Dress
Shoes
Topwear
Topwear
Watches
Fragrance
Accessories
Topwear
Topwear
Topwear
Shoes
Topwear
Topwear
Topwear
Topwear
Shoes
Shoes
Belts
Bottomwear
Shoes
Topwear
Bags
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Topwear
Watches
Belts
Scarves
Topwear
Topwear
Topwear
Topwear
Jewellery
Eyewear
Topwear
Topwear
Topwear
Bags
Topwear
Fragrance
Topwear
Bags
Wallets
Wallets
Bottomwear
Topwear
Topwear
Topwear
Makeup
Topwear
Topwear
Watches
Lips
Shoes
Lips
Topwear
Watches
Headwear
Bags
Topwear
Topwear
Jewellery
Bottomwear
Shoes
Topwear
Eyewear
Shoes
Bags
Wallets
Bags
Bottomwear
Wallets
Bags
Topwear
Top

In [ ]:
pred_labels

['Innerwear',
 'Topwear',
 'Wallets',
 'Shoes',
 'Shoes',
 'Shoes',
 'Topwear',
 'Eyewear',
 'Watches',
 'Shoes',
 'Bags',
 'Topwear',
 'Ties',
 'Innerwear',
 'Socks',
 'Wallets',
 'Watches',
 'Watches',
 'Bottomwear',
 'Shoes',
 'Nails',
 'Topwear',
 'Bottomwear',
 'Belts',
 'Belts',
 'Topwear',
 'Bags',
 'Shoes',
 'Wallets',
 'Topwear',
 'Topwear',
 'Wallets',
 'Topwear',
 'Topwear',
 'Topwear',
 'Eyewear',
 'Watches',
 'Wallets',
 'Watches',
 'Shoes',
 'Topwear',
 'Bottomwear',
 'Bottomwear',
 'Topwear',
 'Watches',
 'Topwear',
 'Shoes',
 'Topwear',
 'Topwear',
 'Bags',
 'Watches',
 'Saree',
 'Innerwear',
 'Topwear',
 'Topwear',
 'Shoes',
 'Wallets',
 'Topwear',
 'Shoes',
 'Watches',
 'Shoes',
 'Shoes',
 'Shoes',
 'Shoes',
 'Dress',
 'Sandal',
 'Topwear',
 'Bottomwear',
 'Topwear',
 'Watches',
 'Accessories',
 'Watches',
 'Topwear',
 'Flip Flops',
 'Watches',
 'Topwear',
 'Topwear',
 'Socks',
 'Topwear',
 'Shoes',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Watches',
 'Wallets

In [ ]:
# pred=[]
# for i in pred_labels:
#   word=""
#    for j in i.split(" "):
#     if j!="and":
#         word=word+(j.capitalize())
#     else:
#         word=word+" "+j+" "
#   pred.append(word)
# pred

['Notebook', 'Book and Shoe', 'Jeans']

In [ ]:
predict=np.stack((test['id'],np.array(pred_labels)),axis=1)
import csv

with open('submission_noisy_attention.csv', mode='w') as csv_file:
    writer = csv.writer(csv_file)
    writer.writerow(['id','category'])
    writer.writerows(predict)

In [ ]:
test.columns

Index(['id', 'gender', 'baseColour', 'season', 'usage',
       'noisyTextDescription'],
      dtype='object')

#5) use decision tree to classify based on 4 categorical features:

In [ ]:
import numpy as np 
import pandas as pds 



In [ ]:
X=train.drop(columns=["id","noisyTextDescription","category"])
X.columns

Index(['gender', 'baseColour', 'season', 'usage'], dtype='object')

In [ ]:
# label encoding the data 

import pandas as pd
from sklearn.preprocessing import OneHotEncoder 
le = LabelEncoder() 
 ### Correct
X_test=test.drop(columns=["id","noisyTextDescription"])
enc = OneHotEncoder(handle_unknown = 'ignore')
enc.fit(X)
X_onehot=enc.transform(X).toarray()
print(enc.transform(X).toarray())

X_test_onehot=enc.transform(X_test).toarray()
print(enc.transform(X_test).toarray())

 



[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]]
[[0. 0. 1. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 1. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


In [ ]:
from sklearn.preprocessing import LabelEncoder 
  
le = LabelEncoder() 
  
Y= le.fit_transform(train['category'])



array([20,  8, 24, ..., 21, 26, 24])

In [ ]:
# AdaBoost Classification

from sklearn import model_selection
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
seed = 7
num_trees = 5000
kfold = model_selection.KFold(n_splits=10, random_state=seed)
model = AdaBoostClassifier(n_estimators=num_trees, random_state=seed)
results = model_selection.cross_val_score(model, X_onehot, Y, cv=kfold)
print(results.mean())

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_split.py:296: FutureWarning: Setting a random_state has no effect since shuffle is False. This will raise an error in 0.24. You should leave random_state to its default (None), or set shuffle=True.
  FutureWarning


NameError: ignored

Combine three csv files

In [ ]:
import pandas as pd

pred_labels=pd.read_csv('submission_labels.csv',header=0)['category']
pred_l=[]
for i in pred_labels:
  rough_pred=i
  for j in list_of_cat:
    if  rough_pred.split(" ")[0]==j.split(" ")[0]:
        rough_pred=j
  pred_l.append(rough_pred) 


NameError: ignored

In [ ]:
pred_image=pd.read_csv('submission_images.csv',header=0)['category']


In [ ]:
pred_noist=pd.read_csv('submission_noist.csv',header=0)['category']


In [ ]:
from collections import Counter
data = Counter(['a','b','a'])
data.most_common()   # Returns all unique items and their counts
#data.most_common(1)  # Returns the highest occurring item
[i for i in data.items()]

[('a', 2), ('b', 1)]

In [ ]:
final_pred=[]
for i,j,k in zip(pred_image,pred_noist,pred_l):
  if i!=j and j!=k:
    final_pred.append(i)
  else:
    lst=[i,j,k]
    final_pred.append(max(set(lst), key=lst.count))
final_pred

['Topwear',
 'Topwear',
 'Belts',
 'Shoes',
 'Innerwear',
 'Shoes',
 'Topwear',
 'Eyewear',
 'Watches',
 'Shoes',
 'Innerwear',
 'Topwear',
 'Ties',
 'Innerwear',
 'Socks',
 'Wallets',
 'Watches',
 'Watches',
 'Bottomwear',
 'Shoes',
 'Nails',
 'Topwear',
 'Bottomwear',
 'Belts',
 'Belts',
 'Topwear',
 'Bags',
 'Shoes',
 'Bags',
 'Topwear',
 'Topwear',
 'Bags',
 'Dress',
 'Bottomwear',
 'Topwear',
 'Eyewear',
 'Watches',
 'Wallets',
 'Watches',
 'Shoes',
 'Topwear',
 'Topwear',
 'Bottomwear',
 'Topwear',
 'Watches',
 'Topwear',
 'Flip Flops',
 'Topwear',
 'Topwear',
 'Bags',
 'Watches',
 'Saree',
 'Innerwear',
 'Topwear',
 'Topwear',
 'Shoes',
 'Wallets',
 'Topwear',
 'Innerwear',
 'Watches',
 'Shoes',
 'Flip Flops',
 'Shoes',
 'Shoes',
 'Dress',
 'Shoes',
 'Topwear',
 'Bottomwear',
 'Topwear',
 'Watches',
 'Topwear',
 'Watches',
 'Topwear',
 'Flip Flops',
 'Watches',
 'Topwear',
 'Topwear',
 'Wallets',
 'Topwear',
 'Shoes',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Watches',
 

In [ ]:
final_pred=[]
for i,j,k in zip(pred_image,pred_noist,pred_l):
    lst=[i,j,k]
    final_pred.append(max(set(lst), key=lst.count))
final_pred

['Topwear',
 'Topwear',
 'Shoes',
 'Shoes',
 'Topwear',
 'Shoes',
 'Topwear',
 'Eyewear',
 'Watches',
 'Shoes',
 'Topwear',
 'Topwear',
 'Ties',
 'Innerwear',
 'Socks',
 'Wallets',
 'Watches',
 'Watches',
 'Bottomwear',
 'Shoes',
 'Nails',
 'Topwear',
 'Bottomwear',
 'Belts',
 'Belts',
 'Topwear',
 'Bags',
 'Shoes',
 'Wallets',
 'Topwear',
 'Topwear',
 'Wallets',
 'Dress',
 'Bottomwear',
 'Topwear',
 'Eyewear',
 'Watches',
 'Wallets',
 'Watches',
 'Shoes',
 'Topwear',
 'Topwear',
 'Bottomwear',
 'Topwear',
 'Shoes',
 'Topwear',
 'Flip Flops',
 'Topwear',
 'Topwear',
 'Bags',
 'Watches',
 'Saree',
 'Innerwear',
 'Topwear',
 'Topwear',
 'Shoes',
 'Wallets',
 'Topwear',
 'Innerwear',
 'Watches',
 'Shoes',
 'Flip Flops',
 'Shoes',
 'Shoes',
 'Dress',
 'Shoes',
 'Topwear',
 'Bottomwear',
 'Topwear',
 'Watches',
 'Shoes',
 'Watches',
 'Shoes',
 'Flip Flops',
 'Watches',
 'Topwear',
 'Topwear',
 'Wallets',
 'Topwear',
 'Shoes',
 'Topwear',
 'Topwear',
 'Topwear',
 'Topwear',
 'Watches',
 'Wal

In [ ]:
count=0
for i,j in zip(pred_image,final_pred):
  if i!=j:
    count+=1
count/len(final_pred)

0.12192528204179767

In [ ]:
cd /content/drive/MyDrive/Colab\ Notebooks/CS680\ Kaggle/

/content/drive/MyDrive/Colab Notebooks/CS680 Kaggle


In [ ]:
predict=np.stack((test['id'],np.array(final_pred)),axis=1)
import csv

with open('combine_submission.csv', mode='w') as csv_file:
    writer = csv.writer(csv_file)
    writer.writerow(['id','category'])
    writer.writerows(predict)